README
NAME
    ToolSet - Load your commonly-used modules in a single import

VERSION
    This documentation describes version 1.00.

SYNOPSIS
    Creating a ToolSet:

         # My/Tools.pm
         package My::Tools;
     
     use base 'ToolSet'; 
     
     ToolSet->use_pragma( 'strict' );
         ToolSet->use_pragma( 'warnings' );
         ToolSet->use_pragma( qw/feature say switch/ ); # perl 5.10
     
     # define exports from other modules
         ToolSet->export(
             'Carp'          => undef,       # get the defaults
             'Scalar::Util'  => 'refaddr',   # or a specific list
         );
     
     # define exports from this module
         our @EXPORT = qw( shout );
         sub shout { print uc shift };
     
     1; # modules must return true

    Using a ToolSet:

         # my_script.pl
     
     use My::Tools;
     
     # strict is on
         # warnings are on
         # Carp and refaddr are imported
     
     carp "We can carp!";
         print refaddr [];
         shout "We can shout, too!";

DESCRIPTION
    ToolSet provides a mechanism for creating logical bundles of modules
    that can be treated as a single, reusable toolset that is imported as
    one. Unlike CPAN bundles, which specify modules to be installed
    together, a toolset specifies modules to be imported together into other
    code.

    ToolSet is designed to be a superclass -- subclasses will specify
    specific modules to bundle. ToolSet supports custom import lists for
    each included module and even supports compile-time pragmas like
    "strict", "warnings" and "feature".

    A ToolSet module does not physically bundle the component modules, but
    rather specifies lists of modules to be used together and import
    specifications for each. By adding the component modules to a
    prerequisites list in a "Makefile.PL" or "Build.PL" for a ToolSet
    subclass, an entire dependency chain can be managed as a single unit
    across scripts or distributions that use the subclass.

INTERFACE
  Setting up
         use base 'ToolSet';

    ToolSet must be used as a base class.

  @EXPORT
         our @EXPORT = qw( shout };
         sub shout { print uc shift }

    Functions defined in the ToolSet subclass can be automatically exported
    during "use()" by listing them in an @EXPORT array.

  "export"
         ToolSet->export(
             'Carp' => undef,                    
             'Scalar::Util' => 'refaddr',
         );

    Specifies packages and arguments to import via "use()". An argument of
    "undef" or the empty string calls "use()" with default imports.
    Arguments should be provided either as a whitespace delimited string or
    in an anonymous array. An empty anonymous array will be treated like
    passing the empty list as an argument to "use()". Here are examples of
    how how specifications will be provided to "use()":

         'Carp' => undef                 # use Carp; 
         'Carp' => q{}                   # use Carp;
         'Carp' => 'carp croak'          # use Carp qw( carp croak );
         'Carp' => [ '!carp', 'croak' ]  # use Carp qw( !carp croak );
         'Carp' => []                    # use Carp ();

    Elements in an array are passed to "use()" as a white-space separated
    list, so elements may not themselves contain spaces or unexpected
    results will occur.

    As of version 1.00, modules may be repeated multiple times. This is
    useful with modules like autouse.

         ToolSet->export(
           autouse => [ 'Carp' => qw(carp croak) ],
           autouse => [ 'Scalar::Util' => qw(refaddr blessed) ],
         );

  "use_pragma"
       ToolSet->use_pragma( 'strict' );         # use strict;
       ToolSet->use_pragma( 'feature', ':5.10' ); # use feature ':5.10';

    Specifies a compile-time pragma to enable and optional arguments to that
    pragma. This must only be used with pragmas that act via the magic $^H
    or "%^H" variables. It must not be used with modules that have other
    side-effects during import() such as exporting functions.

  "no_pragma"
       ToolSet->no_pragma( 'indirect' ); # no indirect;

    Like "use_pragma", but disables a pragma instead.

    If a pragma is specified in both a "use_pragma" and "no_pragma"
    statement, the "use_pragma" will be executed first. This allow turning
    on a pragma with default settings and then disabling some of them.

       ToolSet->use_pragma( 'strict' );
       ToolSet->no_pragma ( 'strict', 'refs' );

  "set_feature" (DEPRECATED)
    See "use_pragma" instead.

  "set_strict" (DEPRECATED)
    See "use_pragma" instead.

  "set_warnings" (DEPRECATED)
    See "use_pragma" instead.

DIAGNOSTICS
    ToolSet will report an error for a module that cannot be found just like
    an ordinary call to "use()" or "require()".

    Additional error messages include:

    *   "Invalid import specification for MODULE" -- an incorrect type was
        provided for the list to be imported (e.g. a hash reference)

    *   "Can't import missing subroutine NAME" -- the named subroutine is
        listed in @EXPORT, but is not defined in the ToolSet subclass

CONFIGURATION AND ENVIRONMENT
    ToolSet requires no configuration files or environment variables.

DEPENDENCIES
    ToolSet requires at least Perl 5.6. ToolSet subclasses will, of course,
    be dependent on any modules they load.

SEE ALSO
    Similar functionality is provided by the Toolkit module, though that
    module requires defining the bundle via text files found within
    directories in "PERL5LIB" and uses source filtering to insert their
    contents as files are compiled.

BUGS
    Please report any bugs or feature using the CPAN Request Tracker. Bugs
    can be submitted through the web interface at
    <http://rt.cpan.org/Dist/Display.html?Queue=ToolSet>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    David A. Golden (DAGOLDEN)

COPYRIGHT AND LICENSE
    Copyright (c) 2005-2008 by David A. Golden. All rights reserved.

    Licensed under Apache License, Version 2.0 (the "License"). You may not
    use this file except in compliance with the License. A copy of the
    License was distributed with this file or you may obtain a copy of the
    License from http://www.apache.org/licenses/LICENSE-2.0

    Files produced as output though the use of this software, shall not be
    considered Derivative Works, but shall be considered the original work
    of the Licensor.

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.



Changes
Revision history for ToolSet

{{$NEXT}}

1.03      2017-05-05 23:15:15-04:00 America/New_York

    [FIXED]

    - Pass tests on newer perls without '.' in @INC.

1.02      2016-03-10 21:01:23-05:00 America/New_York

    [ADDED]

    - Added support for exact control of white space when constructing
      an import line.

1.01      2014-02-27 11:02:19-05:00 America/New_York

    [CHANGED]

    - Updated metadata, repo and bugtracker

    - Switched to managing distribution with Dist::Zilla

1.00 Thu Apr  2 07:05:19 EDT 2009

    - Changed: modules can be repeated in the list to export() to support
      things like autouse.  (Patch from JMELTZER)

0.99 Thu Dec  4 09:17:40 EST 2008

    - Added: use_pragma() and no_pragma() for compile-time pragma support

    - Deprecated: set_strict(), set_warnings() and set_feature() are
      deprecated favor of use_pragma()

    - Changed: switched to the Apache License, version 2.0; (it's clearer,
      relicensable, and is explicit about contributions)

0.14 Thu Nov  6 17:26:10 EST 2008
    - Added: support for feature.pm in Perl 5.10

0.13 Tue May 23 00:09:22 EDT 2006
    - made pod/pod_coverage author tests skip by default
    - removed always-out-of-date version info from Pod
    - moved build_requires to build for better CPANPLUS compatibility

0.12 Sun Dec  4 23:45:35 EST 2005
    - fixed Build.PL generation of Makefile.PL

0.11 Tue Oct 25 06:42:25 EDT 2005
    - adds support for @EXPORT to list functions that ToolSet
      will export by default

0.10 Sat Oct 22 17:14:21 EDT 2005
    - set_strict() added
    - set_warnings() added
    - export() added
    - sample ToolSet::SWC added


t/Util.pm
package t::Util;
use strict;
use warnings;

use Test::More;
use base 'Exporter';

our @EXPORT = qw( cant_ok );

sub cant_ok ($@) {
    my($proto, @methods) = @_;
    my $class = ref $proto || $proto;
    my $tb = Test::More->builder;

    unless( @methods ) {
        my $ok = $tb->ok( 0, "$class->can(...)" );
        $tb->diag('    cant_ok() called with no methods');
        return $ok;
    }

    my @ok = ();
    foreach my $method (@methods) {
        local($!, $@);  # don't interfere with caller's $@
        # eval sometimes resets $!
        eval { $proto->can($method) } && push @ok, $method;
    }

    my $name;
    $name = @methods == 1 ? "$class->cant('$methods[0]')"
    : "$class->cant(...)";

    my $ok = $tb->ok( !@ok, $name );

    $tb->diag(map "    $class->cant('$_') failed\n", @ok);

    return $ok;
}

1;


t/Sample/SWC.pm
package t::Sample::SWC;

use ToolBox::SWC;

carp "We can carp";

1;



t/ToolBox/Null.pm
package t::ToolSet::Null;
use base 'ToolSet';

ToolSet->set_strict(0);
ToolSet->set_warnings(0);

1; # return true


Build.PL
# Copyright (c) 2008 by David Golden. All rights reserved.
# Licensed under Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License was distributed with this file or you may obtain a 
# copy of the License from http://www.apache.org/licenses/LICENSE-2.0

use 5.006;
use strict;
use warnings;
use lib 'inc';
eval "require Pod::WikiDoc";
my $class = $@ ? "Module::Build" : "Module::Build::WikiDoc";
eval "require $class";
my $build = $class->new( 
    module_name => 'ToolSet',
    dist_author => 'David Golden <dagolden@cpan.org>',
    license => 'apache',
    create_readme => 1,
    create_makefile_pl  => 'traditional',
    requires => {
      'perl' => 5.006,
      'base' => 0, 
      'Carp' => 0,
    },
    build_requires => {
      'Exporter' => 0,
      'Test::More' => 0.47, 
    },
    meta_add            => { 
        no_index => { 
            directory => [ qw/ examples inc t /],
        } 
    },
);
$build->create_build_script;



Makefile.PL
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.17;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Load your commonly-used modules in a single import",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "ToolSet",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "ToolSet",
  "PREREQ_PM" => {
    "Carp" => 0,
    "base" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Exporter" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0,
    "lib" => 0
  },
  "VERSION" => "1.04",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Test::More" => 0,
  "base" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/ToolBox.pm
package ToolSet;

use 5.006;
use strict;
use warnings;
use Carp;

our $VERSION = "0.10";

#--------------------------------------------------------------------------#
# package variables
#--------------------------------------------------------------------------#

my %use_strict;
my %use_warnings;
my %exports_of;

#--------------------------------------------------------------------------#
# functions
#--------------------------------------------------------------------------#

sub export {
    my $class = shift;
    my %spec = @_;
    my $caller = caller;
    $exports_of{ $caller } = \%spec;
}

sub import {
    my ($class) = @_;
    my $caller = caller;
    if ( $use_strict{ $class } ) {
        require strict; 
        strict->import;
    }
    if ( $use_warnings{ $class } ) {
        require warnings;
        warnings->import;
    }
    while ( my ( $mod, $request ) = each %{ $exports_of{ $class } } ) {
        my $evaltext;
        if ( ! $request ) {
            $evaltext = "package $caller; use $mod";
        }
        elsif ( ref $request eq 'ARRAY' ) {
            my $args = join( q{ } => @$request );
            $evaltext =  "package $caller; use $mod qw( $args )";
        }
        elsif ( ref( \$request ) eq 'SCALAR' ) {
            $evaltext = "package $caller; use $mod qw( $request )";
        }
        else {
            croak "Invalid import specification for $mod";
        }
        eval $evaltext;
        croak "$@" if $@;
    }
                
}

sub set_strict {
    my ($class, $value) = @_;
    my $caller = caller;
    $use_strict{ $caller } = $value || 0;
}

sub set_warnings {
    my ($class, $value) = @_;
    my $caller = caller;
    $use_warnings{ $caller } = $value || 0;
}


1; # Magic true value required at end of module
__END__

=begin wikidoc

= NAME

ToolSet - Load your commonly-used modules in a single import

= VERSION

This document describes ToolSet version 0.10

= SYNOPSIS

Creating a ToolSet:

    # My/Tools.pm
    package My::Tools;

    use base 'ToolSet'; 
    
    ToolSet->set_strict(1);
    ToolSet->set_warnings(1);

    ToolSet->export(
        'Carp'          => undef,       # get the defaults
        'Scalar::Util'  => 'refaddr',   # or a specific list
    );

    1; # modules must be true

Using a ToolSet:

    # my_script.pl
    
    use My::Tools;
    
    # strict is on
    # warnings are on
    # Carp and refaddr are imported
    
    carp "We can carp!";
    print refaddr []; 
  
= DESCRIPTION

ToolSet provides a mechanism for creating logical bundles of modules that can
be treated as a single, reusable toolset that is imported as one.  Unlike
CPAN bundles, which specify modules to be installed together, a toolset
specifies modules to be imported together into other code.  

ToolSet is designed to be a superclass -- subclasses will specify specific
modules to bundle.  ToolSet supports custom import lists for each included
module and even supports the {strict} and {warnings} pragmas, optionally
enabling those pragmas when the ToolSet subclass is used.

A ToolSet module does not physically bundle the component modules, but rather
specifies lists of modules to be used together and import specifications for
each.  By adding the component modules to a prerequisites list in a
{Makefile.PL} or {Build.PL} for a ToolSet subclass, an entire dependency chain
can be managed as a single unit across scripts or distributions that use the
subclass.

= INTERFACE 

== Setting up

    use base 'ToolSet';
    
ToolSet must be used as a base class.

== {export}

    ToolSet->export(
        'Carp' => undef,                    
        'Scalar::Util' => 'refaddr',
    );

Specifies packages and arguments to import via {use()}.  An argument of {undef}
or the empty string calls {use()} with default imports.  Arguments should be
provided either as a whitespace delimited string or in an anonymous array.  An
empty anonymous array will be treated like passing the empty list as an
argument to {use()}.  Here are examples of how how specifications will be
provided to {use()}:

    'Carp' => undef                 # use Carp; 
    'Carp' => q{}                   # use Carp;
    'Carp' => 'carp croak'          # use Carp qw( carp croak );
    'Carp' => [ '!carp', 'croak' ]  # use Carp qw( !carp croak );
    'Carp' => []                    # use Carp (); 
    
Elements in an array are passed to {use()} as a white-space separated list, so
elements may not themselves contain spaces or unexpected results will occur.

== {set_strict}

  ToolSet->set_strict(1);
  ToolSet->set_strict(0); # default

Determines whether strict will enabled for modules that {use()} this one.

== {set_warnings}

  ToolSet->set_warnings(1);
  ToolSet->set_warnings(0); # default

Determines whether warnings will enabled for modules that {use()} this one.

= DIAGNOSTICS

ToolSet will report an error for a module that cannot be found just like an
ordinary call to {use()} or {require()}.

Additional error messages include:

* {"Invalid import specification for MODULE"} -- an incorrect type was
provided for the list to be imported (e.g. a hash reference)

= CONFIGURATION AND ENVIRONMENT

ToolSet requires no configuration files or environment variables.

= DEPENDENCIES

ToolSet requires at least Perl 5.6.  ToolSet subclasses will, of course, be
dependent on any modules they load.

= INCOMPATIBILITIES

The only pragmas that are explicitly supported are {strict} and {warnings}, as
these have a lexical scope and will not function properly if called from within
the {eval} statement used for importing other modules.  Other pragmas may or
may not work depending on if they have a compile-time or run-time effect.

= SEE ALSO

Similar functionality is provided by the [Toolkit] module, though that 
module requires defining the bundle via text files found within directories
in {PERL5LIB} and uses source filtering to insert their contents as files
are compiled.

= BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests at
{bug-toolset@rt.cpan.org}, or through the web interface at
[http://rt.cpan.org].

When submitting a bug or request, please include a test-file or a patch to an
existing test-file that illustrates the bug or desired feature.

= AUTHOR

David A Golden  {<dagolden@cpan.org>}

= LICENCE AND COPYRIGHT

Copyright (c) 2005, David A Golden {<dagolden@cpan.org>}. All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See the {LICENSE} file included with this
module.

= DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.

=end wikidoc


lib/ToolSet.pm
use 5.006;
use strict;
use warnings;

package ToolSet;
# ABSTRACT: Load your commonly-used modules in a single import

our $VERSION = '1.04';

use Carp;

#--------------------------------------------------------------------------#
# package variables
#--------------------------------------------------------------------------#

my %use_pragmas;
my %no_pragmas;
my %exports_of;

#--------------------------------------------------------------------------#
# functions
#--------------------------------------------------------------------------#

sub export {
    my $class = shift;
    croak "Arguments to export() must be key/value pairs"
      if @_ % 2;
    my @spec   = @_;
    my $caller = caller;
    $exports_of{$caller} = \@spec;
}

sub import {
    my ($class) = @_;
    my $caller = caller;
    if ( $use_pragmas{$class} ) {
        for my $p ( keys %{ $use_pragmas{$class} } ) {
            my $module = $p;
            $module =~ s{::}{/}g;
            $module .= ".pm";
            require $module;
            $p->import( @{ $use_pragmas{$class}{$p} } );
        }
    }
    if ( $no_pragmas{$class} ) {
        for my $p ( keys %{ $no_pragmas{$class} } ) {
            my $module = $p;
            $module =~ s{::}{/}g;
            $module .= ".pm";
            require $module;
            $p->unimport( @{ $no_pragmas{$class}{$p} } );
        }
    }
    my @exports = @{ $exports_of{$class} || [] };
    while (@exports) {
        my ( $mod, $request ) = splice( @exports, 0, 2 );

        my $evaltext;
        if ( !$request ) {
            $evaltext = "package $caller; use $mod";
        }
        elsif ( ref $request eq 'ARRAY' ) {
            my $args = join( q{ } => @$request );
            $evaltext = "package $caller; use $mod qw( $args )";
        }
        elsif ( ref $request eq 'SCALAR' ) {
            my $args = $$request;
            $evaltext = "package $caller; use $mod $args";
        }
        elsif ( ref( \$request ) eq 'SCALAR' ) {
            $evaltext = "package $caller; use $mod qw( $request )";
        }
        else {
            croak "Invalid import specification for $mod";
        }
        eval $evaltext; ## no critic
        croak "$@" if $@;
    }

    # import from a @EXPORT array in the ToolSet subclass
    {
        no strict 'refs'; ## no critic
        for my $fcn ( @{"${class}::EXPORT"} ) {
            my $source = "${class}::${fcn}";
            die "Can't import missing subroutine $source"
              if !defined *{$source}{CODE};
            *{"${caller}::${fcn}"} = \&{$source};
        }
    }
}

sub set_strict {
    my ( $class, $value ) = @_;
    return unless $value;
    my $caller = caller;
    $use_pragmas{$caller}{strict} = [];
}

sub set_warnings {
    my ( $class, $value ) = @_;
    return unless $value;
    my $caller = caller;
    $use_pragmas{$caller}{warnings} = [];
}

sub set_feature {
    my ( $class, @args ) = @_;
    return unless @args;
    my $caller = caller;
    $use_pragmas{$caller}{feature} = [@args];
}

sub use_pragma {
    my ( $class, $pragma, @args ) = @_;
    my $caller = caller;
    $use_pragmas{$caller}{$pragma} = [@args];
}

sub no_pragma {
    my ( $class, $pragma, @args ) = @_;
    my $caller = caller;
    $no_pragmas{$caller}{$pragma} = [@args];
}

1; # Magic true value required at end of module
__END__

=head1 SYNOPSIS

Creating a ToolSet:

    # My/Tools.pm
    package My::Tools;

    use base 'ToolSet';

    ToolSet->use_pragma( 'strict' );
    ToolSet->use_pragma( 'warnings' );
    ToolSet->use_pragma( qw/feature say switch/ ); # perl 5.10

    # define exports from other modules
    ToolSet->export(
        'Carp'          => undef,       # get the defaults
        'Scalar::Util'  => 'refaddr',   # or a specific list
    );

    # define exports from this module
    our @EXPORT = qw( shout );
    sub shout { print uc shift };

    1; # modules must return true

Using a ToolSet:

    # my_script.pl

    use My::Tools;

    # strict is on
    # warnings are on
    # Carp and refaddr are imported

    carp "We can carp!";
    print refaddr [];
    shout "We can shout, too!";

=head1 DESCRIPTION

ToolSet provides a mechanism for creating logical bundles of modules that can
be treated as a single, reusable toolset that is imported as one.  Unlike
CPAN bundles, which specify modules to be installed together, a toolset
specifies modules to be imported together into other code.

Toolset is designed to be a superclass -- subclasses will specify specific
modules to bundle.  ToolSet supports custom import lists for each included
module and even supports compile-time pragmas like C<strict>, C<warnings>
and C<feature>.

A ToolSet module does not physically bundle the component modules, but
rather specifies lists of modules to be used together and import
specifications for each.  By adding the component modules to a
prerequisites list in a C<Makefile.PL> or C<Build.PL> for a ToolSet
subclass, an entire dependency chain can be managed as a single unit across
scripts or distributions that use the subclass.

=head1 INTERFACE

=head2 Setting up

    use base 'ToolSet';

ToolSet must be used as a base class.

=head2 C<@EXPORT>

    our @EXPORT = qw( shout };
    sub shout { print uc shift }

Functions defined in the ToolSet subclass can be automatically exported during
C<use()> by listing them in an C<@EXPORT> array.

=head2 C<export>

    ToolSet->export(
        'Carp' => undef,
        'Scalar::Util' => 'refaddr',
    );

Specifies packages and arguments to import via C<use()>.  An argument of C<undef>
or the empty string calls C<use()> with default imports.  Arguments should be
provided either as a whitespace delimited string or in an anonymous array.  An
empty anonymous array will be treated like passing the empty list as an
argument to C<use()>.  Here are examples of how how specifications will be
provided to C<use()>:

    'Carp' => undef                 # use Carp;
    'Carp' => q{}                   # use Carp;
    'Carp' => 'carp croak'          # use Carp qw( carp croak );
    'Carp' => [ '!carp', 'croak' ]  # use Carp qw( !carp croak );
    'Carp' => []                    # use Carp ();

Elements in an array are passed to C<use()> as a white-space separated
list, so elements may not themselves contain spaces or unexpected results
will occur.  (But read further below about exact whitespace handling.)

As of version 1.00, modules may be repeated multiple times.  This is useful
with modules like L<autouse>.

    ToolSet->export(
      autouse => [ 'Carp' => qw(carp croak) ],
      autouse => [ 'Scalar::Util' => qw(refaddr blessed) ],
    );

As of version 1.02, if you need exact handling of whitespace, pass a
reference to a string and it will be used verbatim:

    'Foo' => \'-foo => { bar => "Some thing with spaces" }'

    #  use Foo -foo => { bar => "Some thing with spaces" }'

=head2 C<use_pragma>

  ToolSet->use_pragma( 'strict' );         # use strict;
  ToolSet->use_pragma( 'feature', ':5.10' ); # use feature ':5.10';

Specifies a compile-time pragma to enable and optional arguments to that
pragma.  This must only be used with pragmas that act via the magic C<$^H> or
C<%^H> variables.  It must not be used with modules that have other side-effects
during import() such as exporting functions.

=head2 C<no_pragma>

  ToolSet->no_pragma( 'indirect' ); # no indirect;

Like C<use_pragma>, but disables a pragma instead.

If a pragma is specified in both a C<use_pragma> and C<no_pragma> statement, the
C<use_pragma> will be executed first.  This allow turning on a pragma with
default settings and then disabling some of them.

  ToolSet->use_pragma( 'strict' );
  ToolSet->no_pragma ( 'strict', 'refs' );

=head2 C<set_feature> (DEPRECATED)

See C<use_pragma> instead.

=head2 C<set_strict> (DEPRECATED)

See C<use_pragma> instead.

=head2 C<set_warnings> (DEPRECATED)

See C<use_pragma> instead.

=head1 DIAGNOSTICS

ToolSet will report an error for a module that cannot be found just like an
ordinary call to C<use()> or C<require()>.

Additional error messages include:

=for :list
* C<"Invalid import specification for MODULE"> -- an incorrect type was
  provided for the list to be imported (e.g. a hash reference)
* C<"Can't import missing subroutine NAME"> -- the named subroutine is
  listed in C<@EXPORT>, but is not defined in the ToolSet subclass

=head1 CONFIGURATION AND ENVIRONMENT

ToolSet requires no configuration files or environment variables.

=head1 DEPENDENCIES

ToolSet requires at least Perl 5.6.  ToolSet subclasses will, of course, be
dependent on any modules they load.

=head1 SEE ALSO

Similar functionality is provided by the L<Toolkit> module, though that
module requires defining the bundle via text files found within directories
in C<PERL5LIB> and uses source filtering to insert their contents as files
are compiled.

=cut
